### PR TITLE
Run Prepare and Initialize from MvxViewModelLoader

### LINKED ISSUE #1972: Initialize not called when manually instantiating an MvxViewModel
To help us fix your issue, please provide the information in the below template. 

> Note: There is often little we can do without a minimal reproducible sample of the issue, so please provide that in a **standalone git repository** and link it here.
>
> Please only file bugs verified on latest version of MvvmCross and _stable_ Xamarin tools

### Steps to reproduce

1. when manually instantiating an MvxViewModel, I use the following code so that methods of the VM life cycle (`Init()`, `Start()`) are called:

```
var mvxViewModelLoader = Mvx.Resolve<IMvxViewModelLoader>();
var vmRequest = MvxViewModelRequest.GetDefaultRequest(typeof(MyViewModel));
_myViewModel = mvxViewModelLoader.LoadViewModel(vmRequest, null) as MyViewModel;
```

This is typically done in a Parent-Child VM setup (e.g. Tab Bars), where the Parent VM has to instantiate the Children VMs.

### Expected behavior
Using the above method, `Initialize()` should be called in the Child VM so that the vm can be inited (e.g. populated via web services).

### Actual behavior
Using the above method, `Initialize()` is not called on the Child VM, (although `Start()` is).

### Configuration

**Version:** 5.0.3

**Platform:** 
- [x] iOS
- [x] Android
- [ ] WPF
- [ ] UWP

### LINKED ISSUE #2105: Feedback: The new Navigation Service and the life cycle it introduces
I spent the day upgrading a relatively new and small project from 4.4.0 to 5.1.1.0. Now I am considering a rollback. 

One thing I am stumbling upon is the way the new navigation service is working. Beginning with https://github.com/MvvmCross/MvvmCross/commit/80ef9b903acec0e541259ae058744892876cd405 the navigation service creates new instances of the targeted view models the moment `Navigate<T>()` is being called.

This is causing me some headaches:

1. The View Model is being created – the constructor is called.
2. The View Model is being initialized by calling `Initialize()`.
3. The View Model is passed to the presenter as part of a `MvxViewModelInstanceRequest`.

Now let's say we want to pass it to an Android Activity. What we need to to, and what the framework is doing when I didn't miss anything, is serializing the actual View Model at hand and passing it along with the Intent as an extra.

That serialization/deserialization was fine (and necessary, of course) as long as the passed data has been "just meta data" about the View Model which should be used along with the targeted View. But now we have not just meta data in our request, but a whole complex View Model, which may even be doing something the moment we're starting the serialization. 

I am not sure about that. But even if I consider it not a problem, I have to make sure that I am passing always the whole full `MvxViewModelInstanceRequest` to my views, and not accidentaly just a `MvxViewModelRequest`. If that happens (and it happened to me today, not sure anymore in which cases exactly), the View Model will be created for a second time, when using the `IMvxViewModelLoader`. And of course it will be initialized for a second time, too.

Which brings me to the second issue: I am not sure about the new `Task Initialize()`. As this is being awaited the moment the navigation to the view model starts, the whole navigation process stops for as long as that initialization lasts.

I don't know any scenario where I want to have my users experience a frozen UI. What I want my users to see is a view opening the moment they press a button. If initialization takes a while, that's fine – but it has to be running in the background. In the UI we can show some kind of a loading indicator and we can offer the option to cancel, e.g. closing the whole view.

tl;dr: 

1. I think creating and initializing the View Model along with the View has been the better approach.
2. I don't see where and why I should use the `Task Initialize()` for actual long-running initialization work that justifies a Task to be returned.

Thomas

### LINKED ISSUE #2167: ViewModel's SaveState/ReloadState and NavigationService
This is a concern for every platform, but maybe Android is the most affected one because if we don't do something, apps will crash. 

Basically the SaveState/ReloadState mechanism is activated when the device goes out of memory, so what we do internally is to provide a way for users to save the state of their ViewModels and then a way to reload it. This is how it goes inside the framework:

1. If there is only one ViewModel going into tombstone (of if many, then the last one): This ViewModel will not be destroyed, it will be saved in a "fast" cache named `IMvxSingleViewModelCache` instead. Then when the View is coming back, the ViewModel is just retrieved from the cache.

2. For the rest of the ViewModels (these are probably part of a backstack): When the Android View is destroyed, each ViewModel gets a chance to save its state by populating a `Bundle` (which contains a `Dictionary<string,string>`) before being disposed. After this process - when the View comes back - the ViewModel is loaded from the ground up, and `ReloadState` is supposed to be called with a `Bundle` that contains the previously saved data.

I have done some testing for this and so far it seems to be working for Activities + new AndroidViewPresenter + MvxNavigationService I didn't test it for fragments yet, as we didn't finish the caching functionality yet.

__BUT__ there is a problem with lifecycle events for ViewModels: As ReloadState is part of the “old” CIRS (Constructor - Init - Reload State - Start), it is called by `IMvxViewModelLocator.RunViewModelLifecycle`, so currently this is what the chain looks like when a ViewModel is coming back to life:
__Init -> ReloadState -> Start.__

Just that. `Prepare` and `Initialize` are not getting called.

This makes sense - in a way - but if these methods are not getting called then the developer can't really recreate the ViewModel. 

That being said, I think it will be really difficult to make a call to `Initialize` when reloading the state, because we are working in a sync context (View lifecycle events).

I have two alternatives in mind right now:
1) Store the entire stack of ViewModels in a cache singleton (instead of the last one opened). This will probably require the developer to be responsible for configuring which ViewModels should be stored in that cache.

2) Leave the process as it is right now. This means that once we finally deprecate CIRS, the process or saving/reloading will only involve `SaveState` -> `ReloadState`. This leaves the responsibility of running the ViewModel initialization to the user.

### LINKED ISSUE #2297: ViewModel Initialize method not called using MvxViewPagerFragmentInfo
When you create a viewpager with MvxViewPagerFragmentInfo, the ViewModel was created, but the Initialize method was not executed.

### Steps to reproduce

1. Create a basic ViewModel with Initialize method.

2. Configure the viewpager in the Android view, inside the OnCreated method:

```
var viewPager = FindViewById<ViewPager>(Resource.Id.simple_view_pager);
var fragments = new List<MvxViewPagerFragmentInfo>
{
	new MvxViewPagerFragmentInfo("Page 1", typeof(FirstFragment), typeof(FirstViewModel)),
	new MvxViewPagerFragmentInfo("Page 2", typeof(SecondFragment), typeof(SecondViewModel)),
	new MvxViewPagerFragmentInfo("Page 3", typeof(ThirdFragment), typeof(ThirdViewModel)),
};

viewPager.Adapter = new MvxCachingFragmentStatePagerAdapter(this, SupportFragmentManager, fragments);
```

### Expected behavior
The ViewModels execute Initialize method when it's instatiated.

### Actual behavior
The ViewModels doesn't execute the Initialize method when it's instatiated.

### Configuration

**Version:** 5.3.0

**Platform:** 
- [ ] iOS
- [x] Android
- [ ] WPF
- [ ] UWP

### PR DESCRIPTION
**Pull request by nmilcoff**
### :sparkles: What kind of change does this PR introduce? (Bug fix, feature, docs update...)
Bug fix / feature.

### :arrow_heading_down: What is the current behavior?
1) Prepare and Initialize are called from MvxNavigationService directly. As a result, when manually loading a ViewModel the new lifecycle is not run. This is a quite annoying problem when using ViewPagers on Android, 
2) SaveState/ReloadState are not really handled when using the new NavigationService. Prepare nor Initialize are called during the process.
3) It is difficult to know when the Initialize task is finished in ViewModels. 
4) When using Uri Navigation, a Navigate call might fail for different reasons but users will always get a NullReferenceException because a null result is not handled inside MvxNavigationService.

### :new: What is the new behavior (if this is a feature change)?
1) Prepare and Initialize are called from MvxViewModelLoader as part of the ViewModel lifecycle 💪 
2) SaveState / ReloadState works pretty much the same as it did in MvvmCross v4: Users need to save all the state they want to preserve using SaveState. Then, after the tombstoning process, a ViewModel will run: (1) Parameterless Prepare -> (2) ReloadState (time to get your data back!) -> (3) Initialize! 🎉 
3) NotifyTask bits were added into the repository (from [Mvvm.Async](https://github.com/StephenCleary/Mvvm.Async)). Now if you want to know when Initialize task finishes, all you have to do is to subscribe to `ViewModel.InitializeTask` property changes! As a bonus, you can think about binding view properties to the task's states (more info in [this](http://nmilcoff.com/2017/07/10/stop-toggling-isbusy-with-notifytask/) blog post). 👂 
4) This is fixed.

Bonus: Added some more Unit Tests for MvxNavigationService ✔️ 

### :boom: Does this PR introduce a breaking change?
Yes. Both `IMvxViewModelLoader` and `IMvxViewModelLocator` interfaces have changed, since two methdos were added. But there are no breaking changes in terms of behavior (ShowViewModel will continue working, but it will also call Prepare and Initialize now).

### :bug: Recommendations for testing
Run TestProjects!

### :memo: Links to relevant issues/docs
**Docs** that need to be updated:
- [x] [ViewModel lifecycle](https://www.mvvmcross.com/documentation/fundamentals/viewmodel-lifecycle?scroll=76): We need to document how to use SaveState/ReloadState.
- [x] NotifyTask usage (doesn't exist yet)
- [x] [App lifecycle](https://www.mvvmcross.com/documentation/fundamentals/app-lifecycle?scroll=189)

**Issues**
- Fixes #2297
- Closes #2167
- Closes #2105 (MvxNavigationService has now a better shape)
- Fixes #1972

### :thinking: Checklist before submitting

- [X] All projects build
- [X] Follows style guide lines ([code style guide](https://github.com/MvvmCross/MvvmCross#code-style-guidelines))
- [x] Relevant documentation was updated ([docs style guide](https://www.mvvmcross.com/documentation/contribute/mvvmcross-docs-style-guide))
- [X] Nuspec files were updated (when applicable)
- [X] Rebased onto current develop

### CODE REVIEW

**Comment by nmilcoff** [comment]
I'm facing [this issue](https://bugzilla.xamarin.com/show_bug.cgi?id=60233) when trying to debug Unit Tests on my machine :( I'd appreciate if anyone could give me a hand with that.

File: MvvmCross/Core/Core/ViewModels/NotifyTask.cs [hunk 1]
```diff
@@ -0,0 +1,299 @@
...
+using System;
+using System.ComponentModel;
+using System.Threading.Tasks;
+
+namespace MvvmCross.Core.ViewModels
+{
+    /// <summary>
+    /// Watches a task and raises property-changed notifications when the task completes.
+    /// </summary>
+    public sealed class NotifyTask : INotifyPropertyChanged
```
Comment (martijn00) [inline MvvmCross/Core/Core/ViewModels/NotifyTask.cs]: Maybe we should name this `MvxNotifyTask`? That would make it a bit more logical to also use outside the new viewmodel loading.

**Comment by martijn00** [comment]
@nmilcoff Can you make AppVeyor happy?

**Comment by nmilcoff** [comment]
Alright, this PR should be ready to be reviewed now. I have added some unit tests to NavigationService and also updated some docs.

**Comment by softlion** [comment]
> (1) Parameterless Prepare -> (2) ReloadState (time to get your data back!) -> (3) Initialize!

What about ViewModel parameters ?
Are they reloaded ?
Or only the parameterless methods are used ?

File: MvvmCross/Core/Core/ViewModels/MvxNotifyTask.cs [hunk 2]
```diff
@@ -0,0 +1,299 @@
+﻿// Honours to Stephen Cleary for this file's code
```
[the remark this answers is not in the document]
Comment (nmilcoff) [reply MvvmCross/Core/Core/ViewModels/MvxNotifyTask.cs]: That's strange, I didn't disable it. Looks like @martijn00 did it, thanks guys

**Comment by nmilcoff** [comment]
@softlion I've updated the doc https://github.com/MvvmCross/MvvmCross/blob/develop/docs/_documentation/fundamentals/viewmodel-lifecycle.md#tombstoning-saving-and-restoring-the-viewmodels-state 🙂

File: MvvmCross/Core/Core/ViewModels/NotifyTask.cs [hunk 3]
```diff
@@ -0,0 +1,299 @@
+﻿// Honours to Stephen Cleary for this file's code
```
[the remark this answers is not in the document]
Comment (nmilcoff) [reply MvvmCross/Core/Core/ViewModels/NotifyTask.cs]: We're are already using some bits under MIT license and I don't think we have included their full license in the repo. I'm no expert in this topic though... if you say / think is necessary then we should do it

**Pull request merged**